### PR TITLE
(enhancement)resources: add argo rollouts as a supported parent resource for chaos

### PR DESCRIPTION
**Pull request by ksatchit**
Signed-off-by: ksatchit <karthik.s@mayadata.io>

**What this PR does / why we need it**:

- Supports chaos on [argo rollouts](https://argoproj.github.io/argo-rollouts/) resource to facilitate chaos as a step in blue-green/canary deployments
- Also fixes a bug in the deploymentconfig list function in resource pkg

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Checklist:**
-   [ ] Fixes #<issue number>
-   [ ] Labelled this PR & related issue with `documentation` tag
-   [ ] PR messages has document related information
-   [ ] Labelled this PR & related issue with `breaking-changes` tag
-   [ ] PR messages has breaking changes related information
-   [ ] Labelled this PR & related issue with `requires-upgrade` tag
-   [ ] PR messages has upgrade related information
-   [ ] Commit has unit tests
-   [ ] Commit has integration tests

### CODE REVIEW

**Comment by ksatchit** [comment]
## Test Rollout

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Rollout
metadata:
  name: rollouts-demo
  annotations:
    litmuschaos.io/chaos: true
  labels:
    app: rollouts-demo
spec:
  replicas: 5
  strategy:
    canary:
      steps:
      - setWeight: 20
      - pause: {}
      - setWeight: 40
      - pause: {duration: 10}
      - setWeight: 60
      - pause: {duration: 10}
      - setWeight: 80
      - pause: {duration: 10}
  revisionHistoryLimit: 2
  selector:
    matchLabels:
      app: rollouts-demo
  template:
    metadata:
      labels:
        app: rollouts-demo
    spec:
      containers:
      - name: rollouts-demo
        image: argoproj/rollouts-demo:blue
        ports:
        - name: http
          containerPort: 8080
          protocol: TCP
        resources:
          requests:
            memory: 32Mi
            cpu: 5m
```

**Comment by ksatchit** [comment]
## Test Rollout (Negative) 

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Rollout
metadata:
  name: rollouts-demo
  annotations:
    litmuschaos.io/chaos: false
  labels:
    app: rollouts-demo
spec:
  replicas: 5
  strategy:
    canary:
      steps:
      - setWeight: 20
      - pause: {}
      - setWeight: 40
      - pause: {duration: 10}
      - setWeight: 60
      - pause: {duration: 10}
      - setWeight: 80
      - pause: {duration: 10}
  revisionHistoryLimit: 2
  selector:
    matchLabels:
      app: rollouts-demo
  template:
    metadata:
      labels:
        app: rollouts-demo
    spec:
      containers:
      - name: rollouts-demo
        image: argoproj/rollouts-demo:blue
        ports:
        - name: http
          containerPort: 8080
          protocol: TCP
        resources:
          requests:
            memory: 32Mi
            cpu: 5m
```


## ChaosEngine Events 

```yaml
Name:         rollouts-chaos
Namespace:    default
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"litmuschaos.io/v1alpha1","kind":"ChaosEngine","metadata":{"annotations":{},"name":"rollouts-chaos","namespace":"default"},"...
API Version:  litmuschaos.io/v1alpha1
Kind:         ChaosEngine
Metadata:
  Creation Timestamp:  2020-11-14T19:10:41Z
  Generation:          4
  Resource Version:    5917488
  Self Link:           /apis/litmuschaos.io/v1alpha1/namespaces/default/chaosengines/rollouts-chaos
  UID:                 2ec4771b-c9cf-400d-b482-5061bfb1133d
Spec:
  Annotation Check:  true
  Appinfo:
    Appkind:              rollout
    Applabel:             app=rollouts-demo
    Appns:                default
  Chaos Service Account:  pod-delete-sa
  Components:
    Runner:
  Engine State:  stop
  Experiments:
    Name:  pod-delete
    Spec:
      Components:
        Env:
          Name:   TOTAL_CHAOS_DURATION
          Value:  30
          Name:   CHAOS_INTERVAL
          Value:  10
          Name:   FORCE
          Value:  false
        Resources:
        Status Check Timeouts:
      Rank:             0
  Job Clean Up Policy:  retain
Status:
  Engine Status:  stopped
  Experiments:    <nil>
Events:
  Type     Reason                         Age   From            Message
  ----     ------                         ----  ----            -------
  Normal   ChaosEngineInitialized         101s  chaos-operator  rollouts-chaos-runner created successfully
  Warning  ChaosResourcesOperationFailed  101s  chaos-operator  Unable to filter app by specified info
  Warning  ChaosEngineStopped             100s  chaos-operator  Chaos stopped due to failed app identification
```

## Chaos Operator Logs 

```
{"level":"info","ts":1605381041.0537698,"logger":"controller_chaosengine","msg":"Reconciling ChaosEngine","Request.Namespace":"default","Request.Name":"rollouts-chaos"}
{"level":"info","ts":1605381041.0629547,"logger":"controller_chaosengine","msg":"App key derived from chaosengine is ","appLabelKey":"app"}
{"level":"info","ts":1605381041.062991,"logger":"controller_chaosengine","msg":"App Label derived from Chaosengine is ","appLabelValue":"rollouts-demo"}
{"level":"info","ts":1605381041.0629978,"logger":"controller_chaosengine","msg":"App NS derived from Chaosengine is ","appNamespace":"default"}
{"level":"info","ts":1605381041.0630028,"logger":"controller_chaosengine","msg":"Exp list derived from chaosengine is ","appExpirements":["pod-delete"]}
{"level":"info","ts":1605381041.06301,"logger":"controller_chaosengine","msg":"Monitoring Status derived from chaosengine is","monitoringStatus":false}
{"level":"info","ts":1605381041.0630162,"logger":"controller_chaosengine","msg":"Runner image derived from chaosengine is","runnerImage":"litmuschaos/chaos-runner:1.9.1"}
{"level":"info","ts":1605381041.0630217,"logger":"controller_chaosengine","msg":"Annotation check is ","annotationCheck":"true"}
{"level":"info","ts":1605381041.0670328,"logger":"controller_chaosengine","msg":"Annotation check failed with","error:":"resource type 'rollout', err: no argo rollout chaos-candidate found"}
```

File: pkg/controller/resource/argorollouts.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("error while listing argo rollouts with matching labels %s", engine.Instance.Spec.Appinfo.Applabel)
+	}
+	if len(rolloutList.Items) == 0 {
+		return nil, fmt.Errorf("no argo rollouts with matching labels %s", engine.Instance.Spec.Appinfo.Applabel)
+	}
+	return rolloutList, err
+}
+
+// This will check and count the total chaos enabled application
```
Comment (ispeakc0de) [inline pkg/controller/resource/argorollouts.go:58]: Go linter recommend the comments started with the function name i.e,  `checkForChaosEnabledRollout ...`

File: pkg/controller/resource/argorollouts.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	for _, rollout := range rolloutList.Items {
+		engine.AppName = rollout.GetName()
+		engine.AppUUID = rollout.GetUID()
+		annotationValue := rollout.GetAnnotations()[ChaosAnnotationKey]
+		chaosEnabledRollout = CountTotalChaosEnabled(annotationValue, chaosEnabledRollout)
+		if chaosEnabledRollout > 1 {
+			return engine, chaosEnabledRollout, errors.New("too many argo rollouts with specified label are annotated for chaos, either provide unique labels or annotate only desired rollout for chaos")
+		}
+	}
+	return engine, chaosEnabledRollout, nil
```
Comment (ispeakc0de) [inline pkg/controller/resource/argorollouts.go:71]: engine is call by reference, do we really need to return this?

Comment (ksatchit) [reply pkg/controller/resource/argorollouts.go:71]: Yes, saw this. Seems this is needed on all resources. Will address this in a separate PR

**Review by ispeakc0de — APPROVED** [review]

**Pull request merged**